completions/fish/wl-paste.fish

```diff
@@ -0,0 +1,44 @@
+function __wayland_seats --description 'Print all wayland seats'
+    if type -q weston-info
+        weston-info 2>/dev/null | sed -n '/wl_seat/{n;s/\s*name: //;p}'
+    else if type -q loginctl
+        loginctl --no-legend --no-pager list-seats 2>/dev/null
+    else
+        # Fallback seat
+        echo "seat0"    
+    end
+end
+
+function __wl_paste_types --description 'Print types with context'
+    set -l type (commandline -o)
+
+    for i in (seq (count $type))
+        if [ "$type[$i]" = '--primary' ]; or [ "$type[$i]" = '-p' ]
+            set clip "primary"
+        else if [ "$type[$i]" = '--seat' ]; or [ "$type[$i]" = '-s' ]
+            set seat $type[(math $i + 1)]
+        end
+    end
+
+    # Note fish does not handle passing unset variables
+    # to commands well, thus setting clip to "--primary" and passing
+    # that to wl-paste wont work, so if statements are used instead
+    if test -n "$seat"; and test -n "$clip"
+        wl-paste --seat "$seat" -p -l
+    else if test -n "$seat"
+        wl-paste --seat "$seat" -l
+    else if test -n "$clip"
+        wl-paste -p -l
+    else
+        wl-paste -l
+    end
+end
+
+complete -c wl-paste -s h -l help -d 'Display a help message.'
+complete -c wl-paste -s v -l version -d 'Display version info.'
+complete -c wl-paste -s n -l no-newline -d 'Do not append a newline character.'
+complete -c wl-paste -s l -l list-types -d 'Instead of pasting, list the offered types.'
+complete -c wl-paste -s p -l primary -d 'Use the "primary" clipboard.'
+complete -c wl-paste -s w -l watch -r -d 'Run a command each time the selection changes.'
```

Comment:
This should complete as a subcommand — that is, complete installed executables, and then recursively complete their options. As far as I can see, `-r` stands for *required* argument, not for *recursively* complete a subcommand. It seems you need to use `-w` here.

Reply:
I'm not entirely clear on what you're asking here. Do you want to offer all the commands in $PATH? Using `-w` makes wl-paste inherit all the completions of the given command not recursively complete for the subcommand.

Reply:
Yes, we want `wl-paste` to inherit other command's completions in this case.

For example, if I type

`wl-paste --primary --watch rm --f<TAB><TAB>`

I would expect it to complete `orce` (for `--force`), because that's what it does when you type

`rm --f<TAB><TAB>`

Reply:
Looking at the fish source `complete` seems to only use the first argument on the command line to apply completions to [link](https://github.com/fish-shell/fish-shell/blob/7d486c121ab6bda59bcefd9ffd365576c40e7484/src/builtin_complete.cpp#L116). So I don't think we can offer completions for only the subcommand, but only pass `--watch` command completions to `wl-paste` via  `--wrap`; Unfortunately I haven't  found a good way to do this calling `complete -c wl-paste --wrap rm` causes wl-paste to inherit `rm's` completions until the `wl-paste.fish` file is re-sourced.

Reply:
Yeah sorry, I got confused, `-w` seems to be for making aliases, not subcommands (the docs could be more clear on this though). `__fish_complete_subcommand` is what you need here.